### PR TITLE
[nan-347] cancel script implementation

### PR DESCRIPTION
**Pull request by khaliqgant**
## Describe your changes
Allows a script to be cancelled from the UI by forking the process in the runner and keeping the child in memory to later cancel it if desired.

## Issue ticket number and link
NAN-347

## Checklist before requesting a review (skip if just adding/editing APIs & templates)
- [ ] I added tests, otherwise the reason is: 
- [ ] I added observability, otherwise the reason is:
- [ ] I added analytics, otherwise the reason is:

### CODE REVIEW

File: packages/jobs/lib/activities.ts [hunk 1]
```diff
@@ -407,3 +407,36 @@ export async function reportFailure(
...
+        debug: Boolean(debug)
+    });
+
+    if ('syncJobId' in workflowArguments) {
+        await updateSyncJobStatus(workflowArguments.syncJobId, SyncStatus.STOPPED);
+    } else {
+        await updateLatestJobSyncStatus(workflowArguments.syncId, SyncStatus.STOPPED);
+    }
+
+    await syncRun.cancel();
```
Comment (TBonnin) [inline packages/jobs/lib/activities.ts]: what happen if the cancel or any other awaited function fails

File: packages/shared/lib/services/nango-config.service.ts [hunk 2]
```diff
@@ -475,7 +474,7 @@ export function convertV2ConfigObject(config: NangoConfigV2, showMessages = fals
     return { success: true, error: null, response: output };
 }
 
-export function getOffset(interval: string, date: Date): number {
+export function getOffset(interval: StringValue, date: Date): number {
```
Comment (TBonnin) [inline packages/shared/lib/services/nango-config.service.ts:482]: why `StringValue`?

File: packages/shared/lib/services/sync/run.service.ts [hunk 3]
```diff
@@ -325,7 +331,12 @@ export default class SyncRun {
                     const message = `The integration was run but there was a problem in retrieving the results from the script "${this.syncName}"${
                         syncData?.version ? ` version: ${syncData.version}` : ''
                     }`;
-                    await this.reportFailureForResults(message);
+
+                    if (error.type === 'script_cancelled') {
+                        await this.reportFailureForResults(error.message);
+                    } else {
+                        await this.reportFailureForResults(message);
```
Comment (TBonnin) [inline packages/shared/lib/services/sync/run.service.ts:338]: could this be more generic: `await this.reportFailureForResults(error?.message || message);`

**Review by TBonnin — COMMENTED** [review]
made small comments. looks good otherwise

**Review by TBonnin — APPROVED** [review]

File: packages/jobs/lib/integration.service.ts [hunk 4]
```diff
@@ -10,18 +10,54 @@ import {
...
+        } else {
+            if (activityLogId && environmentId) {
+                await createActivityLogMessage({
+                    level: 'error',
+                    environment_id: environmentId as number,
+                    activity_log_id: activityLogId as number,
+                    content: `Failed to cancel script`,
+                    timestamp: Date.now()
+                });
+            }
```
Comment (bodinsamuel) [inline packages/jobs/lib/integration.service.ts]: I think with the if you are covered already
```suggestion
            if (activityLogId && environmentId) {
                await createActivityLogMessage({
                    level: 'error',
                    environment_id: environmentId,
                    activity_log_id: activityLogId,
                    content: `Failed to cancel script`,
                    timestamp: Date.now()
                });
            }
```

File: packages/jobs/lib/workflows.ts [hunk 5]
```diff
@@ -19,6 +19,11 @@ export async function initialSync(args: InitialSyncArgs): Promise<boolean | obje
     try {
         return await routeSync(args);
     } catch (e: any) {
+        if (isCancellation(e)) {
+            await CancellationScope.nonCancellable(() => cancelActivity(args));
+
+            return false;
+        }
```
Comment (bodinsamuel) [inline packages/jobs/lib/workflows.ts:26]: What happen if that call fails? (not sure it can really just asking)

File: packages/jobs/package.json [hunk 6]
```diff
@@ -20,10 +20,10 @@
...
         "@octokit/plugin-throttling": "^7.0.0",
         "@octokit/rest": "^20.0.1",
-        "@temporalio/activity": "^1.5.2",
-        "@temporalio/client": "^1.5.2",
-        "@temporalio/worker": "^1.5.2",
-        "@temporalio/workflow": "^1.5.2",
+        "@temporalio/activity": "^1.9.1",
+        "@temporalio/client": "^1.9.1",
+        "@temporalio/worker": "^1.9.1",
+        "@temporalio/workflow": "^1.9.1",
```
Comment (bodinsamuel) [inline packages/jobs/package.json:26]: Can you also upgrade shared and server?

File: packages/shared/lib/services/nango-config.service.ts [hunk 7]
```diff
@@ -497,43 +496,43 @@ export function getOffset(interval: string, date: Date): number {
  * and then 1636 etc. The offset should be based on the interval and should never be
  * greater than the interval
  */
-export function getInterval(runs: string, date: Date): ServiceResponse<{ interval: string; offset: number }> {
+export function getInterval(runs: string, date: Date): ServiceResponse<{ interval: StringValue; offset: number }> {
     if (runs === 'every half day') {
-        const response = { interval: '12h', offset: getOffset('12h', date) };
+        const response = { interval: '12h' as StringValue, offset: getOffset('12h', date) };
```
Comment (bodinsamuel) [inline packages/shared/lib/services/nango-config.service.ts]: I would create a type instead of using `as` (or return directly); this keyword will cast the string even tho it's not compatible. e.g: `'foobar' as StringValue` will work but should not

```ts
interface GetIntervalRes { interval: StringValue; offset: number }
export function getInterval(runs: string, date: Date): ServiceResponse<GetIntervalRes> {
   const response: GetIntervalRes = { interval: '12h', offset: getOffset('12h', date) };
   return { success: true, error: null, response };
}
```
 
alternative with type inference

```ts
export function getInterval(runs: string, date: Date): ServiceResponse<{ interval: StringValue; offset: number }> {
   return { success: true, error: null, response: { interval: '12h', offset: getOffset('12h', date) } };
}
```

File: packages/shared/lib/services/sync/job.service.ts [hunk 8]
```diff
@@ -148,10 +148,10 @@ export const isSyncJobRunning = async (sync_id: string): Promise<Pick<SyncJob, '
             deleted: false,
             status: SyncStatus.RUNNING
         })
-        .first('id', 'job_id', 'run_id');
+        .orderBy('created_at', 'desc');
```
Comment (bodinsamuel) [inline packages/shared/lib/services/sync/job.service.ts]: Should we keep the limit?

```suggestion
        .orderBy('created_at', 'desc').limit(1);
```

File: packages/runner/lib/server.ts [hunk 9]
```diff
@@ -59,6 +61,14 @@ function runProcedure() {
         });
 }
 
+function cancelProcedure() {
+    return publicProcedure
+        .input((input) => input as { syncId: string })
+        .mutation(async ({ input }) => {
+            return await kill(input.syncId);
```
Comment (bodinsamuel) [inline packages/runner/lib/server.ts]:
```suggestion
            return kill(input.syncId);
```

File: packages/runner/lib/client.unit.test.ts [hunk 10]
```diff
@@ -49,4 +49,9 @@ describe('Runner client', () => {
         const run = client.run.mutate({ nangoProps, isInvokedImmediately, isWebhook, code: jsCode });
         await expect(run).resolves.toEqual([1, 2, 3]);
     });
+
+    it('should cancel a sync', async () => {
+        const result = await client.cancel.mutate({ syncId: 'sync-id' });
+        expect(result).toEqual({ ok: true, res: 'cancelled' });
```
Comment (bodinsamuel) [inline packages/runner/lib/client.unit.test.ts]: ~I'm surprised it works just a random syncId~
nevermind it cancels the previous test, might be cleaner to not depends on other test otherwise it will be hard to follow and can't be run independently for reproduction

**Review by bodinsamuel — COMMENTED** [review]
It works 👌🏻 
Would be cool to have the button directly in the activity tabs but that's a details.

I just had one additional question, why when it's cancelled I don't get any logs in the activity? 
Screenshot 2024-02-13 at 14 42 33

File: packages/shared/lib/services/sync/run.service.ts (hunk 3 above)
Comment (khaliqgant) [reply packages/shared/lib/services/sync/run.service.ts:338]: I don't want to change the behavior here too much. I believe a NangoError will have a message but we don't necessarily want to expose that to the user. Error messaging should be cleaned up but I think that is out of scope of this work.

File: packages/jobs/lib/workflows.ts (hunk 5 above)
Comment (khaliqgant) [reply packages/jobs/lib/workflows.ts:26]: It would just not cancel correctly and the workflow would continue.

**Comment by khaliqgant** [comment]
> I just had one additional question, why when it's cancelled I don't get any logs in the activity?

It shows on the original sync activity logs:

image

File: packages/shared/lib/services/nango-config.service.ts (hunk 2 above)
Comment (khaliqgant) [reply packages/shared/lib/services/nango-config.service.ts:482]: Temporal expects it now with the latest versions

File: packages/jobs/lib/activities.ts (hunk 1 above)
Comment (khaliqgant) [reply packages/jobs/lib/activities.ts]: Updated to be in a try catch

File: packages/jobs/lib/integration.service.ts (hunk 4 above)
Comment (khaliqgant) [reply packages/jobs/lib/integration.service.ts]: Thanks

File: packages/jobs/package.json (hunk 6 above)
Comment (khaliqgant) [reply packages/jobs/package.json:26]: Updated for shared and the server wasn't using anymore so removed there. Thanks

File: packages/shared/lib/services/nango-config.service.ts (hunk 7 above)
Comment (khaliqgant) [reply packages/shared/lib/services/nango-config.service.ts]: Thanks, updated.

File: packages/shared/lib/services/sync/job.service.ts (hunk 8 above)
Comment (khaliqgant) [reply packages/shared/lib/services/sync/job.service.ts]: Updated, thanks

File: packages/runner/lib/server.ts (hunk 9 above)
Comment (khaliqgant) [reply packages/runner/lib/server.ts]: Updated, thanks

File: packages/runner/lib/client.unit.test.ts (hunk 10 above)
Comment (khaliqgant) [reply packages/runner/lib/client.unit.test.ts]: Kept this for now. This test suite needs to love after this update so will circle back to this

File: packages/runner/lib/client.unit.test.ts [hunk 11]
```diff
@@ -47,6 +47,13 @@ describe('Runner client', () => {
         const isInvokedImmediately = false;
         const isWebhook = false;
         const run = client.run.mutate({ nangoProps, isInvokedImmediately, isWebhook, code: jsCode });
-        await expect(run).resolves.toEqual([1, 2, 3]);
+        console.log(run);
+        // TODO
+        //await expect(run).resolves.toEqual([1, 2, 3]);
```
Comment (khaliqgant) [inline packages/runner/lib/client.unit.test.ts]: @TBonnin are you OK with this for now? I'll try and circle back to this soon so this test can be restored.

File: packages/jobs/lib/workflows.ts [hunk 12]
```diff
@@ -19,6 +19,11 @@ export async function initialSync(args: InitialSyncArgs): Promise<boolean | obje
     try {
         return await routeSync(args);
     } catch (e: any) {
+        if (isCancellation(e)) {
+            await CancellationScope.nonCancellable(() => cancelActivity(args));
+
+            return false;
```
Comment (khaliqgant) [inline packages/jobs/lib/workflows.ts:25]: Make this a consistent return as below

File: packages/runner/lib/client.unit.test.ts (hunk 11 above)
Comment (TBonnin) [reply packages/runner/lib/client.unit.test.ts]: can we at least test the `child.exec(...)` function?

File: packages/shared/lib/services/nango-config.service.ts [hunk 13]
```diff
@@ -546,6 +550,7 @@ export function getInterval(runs: string, date: Date): ServiceResponse<{ interva
     }
```
Comment (bodinsamuel) [inline packages/shared/lib/services/nango-config.service.ts:550]: Not related to your pr, just noticed this condition should come before the previous one

**Review by bodinsamuel — APPROVED** [review]
Massive change 🙏🏻 
If you manage to make the test work that would be great but not a blocker at this point

File: packages/jobs/lib/workflows.ts (hunk 5 above)
Comment (bodinsamuel) [reply packages/jobs/lib/workflows.ts:26]: yeah sorry I miss spoke, I meant since it's not in a try/catch will the error bubble and crashes (or something else)?

File: packages/shared/lib/services/nango-config.service.ts (hunk 13 above)
Comment (khaliqgant) [reply packages/shared/lib/services/nango-config.service.ts:550]: Thanks, updated with https://github.com/NangoHQ/nango/pull/1662/commits/d887a42ba7e4302881058ff7f52bcb2dc03b056a

File: packages/runner/lib/client.unit.test.ts [hunk 14]
```diff
@@ -46,14 +47,8 @@ describe('Runner client', () => {
...
-        console.log(run);
-        // TODO
-        //await expect(run).resolves.toEqual([1, 2, 3]);
-    });
 
-    it('should cancel a sync', async () => {
-        const result = await client.cancel.mutate({ syncId: 'sync-id' });
-        expect(result).toEqual({ ok: true, res: 'cancelled' });
+        const run = await exec(nangoProps, isInvokedImmediately, isWebhook, jsCode);
+        expect(run).toEqual([1, 2, 3]);
```
Comment (TBonnin) [inline packages/runner/lib/client.unit.test.ts:58]: keep a TODO to revert back to testing the entire call to `runner.run` but otherwise great

Comment (khaliqgant) [reply packages/runner/lib/client.unit.test.ts:58]: Thanks, added with https://github.com/NangoHQ/nango/pull/1662/commits/77fb37ab0ef15d8728f8564b2c7f3422c87c4a55
